src/dma.rs

```diff
@@ -0,0 +1,584 @@
+//! # Direct Memory Access
+#![allow(dead_code)]
+
+use core::{
+    marker::PhantomData,
+    sync::atomic::{compiler_fence, Ordering},
+};
+use embedded_dma::{StaticReadBuffer, StaticWriteBuffer};
+
+use crate::rcu::Rcu;
+
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Error {
+    Overrun,
+}
+
+pub enum Event {
+    HalfTransfer,
+    TransferComplete,
+}
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum Half {
+    First,
+    Second,
+}
+
+pub enum Direction {
+    PeripherialToMemory,
+    MemoryToPeripherial,
+    MemoryToMemory,
+}
+
+pub struct CircBuffer<BUFFER, PAYLOAD>
+where
+    BUFFER: 'static,
+{
+    buffer: &'static mut [BUFFER; 2],
+    payload: PAYLOAD,
+    readable_half: Half,
+}
+
+impl<BUFFER, PAYLOAD> CircBuffer<BUFFER, PAYLOAD>
+where
+    &'static mut [BUFFER; 2]: StaticWriteBuffer,
+    BUFFER: 'static,
+{
+    pub(crate) fn new(buf: &'static mut [BUFFER; 2], payload: PAYLOAD) -> Self {
+        CircBuffer {
+            buffer: buf,
+            payload,
+            readable_half: Half::Second,
+        }
+    }
+}
+
+pub trait DmaExt {
+    type Channels;
+
+    fn split(self, rcu: &mut Rcu) -> Self::Channels;
+}
+
+pub trait TransferPayload {
+    fn start(&mut self);
+    fn stop(&mut self);
+}
+
+pub struct Transfer<MODE, BUFFER, PAYLOAD>
+where
+    PAYLOAD: TransferPayload,
+{
+    _mode: PhantomData<MODE>,
+    buffer: BUFFER,
+    payload: PAYLOAD,
+}
+
+impl<BUFFER, PAYLOAD> Transfer<R, BUFFER, PAYLOAD>
+where
+    PAYLOAD: TransferPayload,
+{
+    pub(crate) fn r(buffer: BUFFER, payload: PAYLOAD) -> Self {
+        Transfer {
+            _mode: PhantomData,
+            buffer,
+            payload,
+        }
+    }
+}
+
+impl<BUFFER, PAYLOAD> Transfer<W, BUFFER, PAYLOAD>
+where
+    PAYLOAD: TransferPayload,
+{
+    pub(crate) fn w(buffer: BUFFER, payload: PAYLOAD) -> Self {
+        Transfer {
+            _mode: PhantomData,
+            buffer,
+            payload,
+        }
+    }
+}
+
+impl<MODE, BUFFER, PAYLOAD> Drop for Transfer<MODE, BUFFER, PAYLOAD>
+where
+    PAYLOAD: TransferPayload,
+{
+    fn drop(&mut self) {
+        self.payload.stop();
+        compiler_fence(Ordering::SeqCst);
+    }
+}
+
+/// Read transfer
+pub struct R;
+
+/// Write transfer
+pub struct W;
+
+macro_rules! dma {
+    ($($DMAX:ident: ($dmaX:ident, {
+        $($CX:ident: (
+            $CTLX:ident,
+            $CNTX:ident,
+            $PADDRX:ident,
+            $MADDRX:ident,
+            $ctlX:ident,
+            $cntX:ident,
+            $paddrX:ident,
+            $maddrX:ident,
+            $errifX:ident,
+            $htfifX:ident,
+            $ftfifX:ident,
+            $errifcX:ident,
```

Comment:
`$errifX` and `$errifcX` are not used in the code. Maybe they should be removed?